[mlir][linalg] Add e2e test for linalg.mmt4d + pack/unpack

banach-space opened the pull request:
This is a follow-up for #81790. This patch basically extends:

  * test/Integration/Dialect/Linalg/CPU/mmt4d.mlir

with pack/unpack ops so that to overall computation is a matrix
multiplication (as opposed to linalg.mmt4d). For comparison (and to make
it easier to verify correctness), linalg.matmul is also included in the
test.

banach-space committed:
[mlir][linalg] Add e2e test for linalg.mmt4d

This is a follow-up for #81790. This patch basically extends:

  * test/Integration/Dialect/Linalg/CPU/mmt4d.mlir

with pack/unpack ops so that to overall computation is a matrix
multiplication (as opposed to linalg.mmt4d). For comparison (and to make
it easier to verify correctness), linalg.matmul is also included in the
test.

c-rhodes commented on mlir/test/Integration/Dialect/Linalg/CPU/pack-unpack-mmt4d.mlir:
```diff
@@ -0,0 +1,174 @@
+  %C_pack_empty = tensor.empty() : tensor<2x2x8x8xi32>
+
+  // Pack matrices
+  %A_pack = tensor.pack %A padding_value(%zero : i32) outer_dims_perm = [0, 1] inner_dims_pos = [0, 1] inner_tiles = [8, 1] into %A_pack_empty : tensor<7x16xi32> -> tensor<2x16x8x1xi32>
```
outer_dims_perm isn't doing anything here or for `%C_pack` / `%C_out_unpack`
```suggestion
  %A_pack = tensor.pack %A padding_value(%zero : i32) inner_dims_pos = [0, 1] inner_tiles = [8, 1] into %A_pack_empty : tensor<7x16xi32> -> tensor<2x16x8x1xi32>
```

c-rhodes commented on mlir/test/Integration/Dialect/Linalg/CPU/pack-unpack-mmt4d.mlir:
```diff
@@ -0,0 +1,174 @@
+  // Pack matrices
+  %A_pack = tensor.pack %A padding_value(%zero : i32) outer_dims_perm = [0, 1] inner_dims_pos = [0, 1] inner_tiles = [8, 1] into %A_pack_empty : tensor<7x16xi32> -> tensor<2x16x8x1xi32>
+  %B_pack = tensor.pack %B padding_value(%zero : i32) outer_dims_perm = [1, 0] inner_dims_pos = [1, 0] inner_tiles = [8, 1] into %B_pack_empty : tensor<16x13xi32> -> tensor<2x16x8x1xi32>
+  %C_pack = tensor.pack %C padding_value(%zero : i32) outer_dims_perm = [0, 1] inner_dims_pos = [0, 1] inner_tiles = [8, 8] into %C_pack_empty : tensor<7x13xi32> -> tensor<2x2x8x8xi32>
```
is this a realistic packing? The second 8x8 isn't used at all,
```
Unranked Memref base@ = 0xaaaaee4866c0 rank = 4 offset = 0 sizes = [2, 2, 8, 8] strides = [128, 64, 8, 1] data =
[[[[1,     8,     15,     22,     29,     36,     43,     50],
   [2,     9,     16,     23,     30,     37,     44,     51],
   [3,     10,     17,     24,     31,     38,     45,     52],
   [4,     11,     18,     25,     32,     39,     46,     53],
   [5,     12,     19,     26,     33,     40,     47,     54],
   [6,     13,     20,     27,     34,     41,     48,     55],
   [7,     14,     21,     28,     35,     42,     49,     56],
   [0,     0,     0,     0,     0,     0,     0,     0]],
  [[57,     64,     71,     78,     85,     0,     0,     0],
   [58,     65,     72,     79,     86,     0,     0,     0],
   [59,     66,     73,     80,     87,     0,     0,     0],
   [60,     67,     74,     81,     88,     0,     0,     0],
   [61,     68,     75,     82,     89,     0,     0,     0],
   [62,     69,     76,     83,     90,     0,     0,     0],
   [63,     70,     77,     84,     91,     0,     0,     0],
   [0,     0,     0,     0,     0,     0,     0,     0]]],
 [[[0,     0,     0,     0,     0,     0,     0,     0],
   [0,     0,     0,     0,     0,     0,     0,     0],
   [0,     0,     0,     0,     0,     0,     0,     0],
   [0,     0,     0,     0,     0,     0,     0,     0],
   [0,     0,     0,     0,     0,     0,     0,     0],
   [0,     0,     0,     0,     0,     0,     0,     0],
   [0,     0,     0,     0,     0,     0,     0,     0],
   [0,     0,     0,     0,     0,     0,     0,     0]],
  [[0,     0,     0,     0,     0,     0,     0,     0],
   [0,     0,     0,     0,     0,     0,     0,     0],
   [0,     0,     0,     0,     0,     0,     0,     0],
   [0,     0,     0,     0,     0,     0,     0,     0],
   [0,     0,     0,     0,     0,     0,     0,     0],
   [0,     0,     0,     0,     0,     0,     0,     0],
   [0,     0,     0,     0,     0,     0,     0,     0],
   [0,     0,     0,     0,     0,     0,     0,     0]]]]
```

banach-space replied:
Thanks for checking :) 

> is this a realistic packing

I took these sizes from IREE - so in this sense they are realistic. This is not really to demonstrate anything optimal - it's more about the compilation flow.

c-rhodes reviewed (commented):
thanks for patch Andrzej, I've left some initial comments but I still need to digest what this is doing, dont fully understand it yet

banach-space commented on mlir/test/Integration/Dialect/Linalg/CPU/pack-unpack-mmt4d.mlir:
```diff
@@ -0,0 +1,174 @@
+// DEFINE: %{compile} =  mlir-opt %s \
+// DEFINE:    -transform-interpreter -test-transform-dialect-erase-schedule \
+// DEFINE:    -one-shot-bufferize -func-bufferize -cse -canonicalize -convert-vector-to-scf -test-lower-to-llvm -o %t
```
Thanks!

banach-space committed:
fixup! [mlir][linalg] Add e2e test for linalg.mmt4d

c-rhodes commented on mlir/test/Integration/Dialect/Linalg/CPU/pack-unpack-mmt4d.mlir:
```diff
@@ -0,0 +1,174 @@
+  %B_pack_empty = tensor.empty() : tensor<2x16x8x1xi32>
+  %C_pack_empty = tensor.empty() : tensor<2x2x8x8xi32>
+
+  // Pack matrices
```
Given this is the key part of what's going on here I'd find a more descriptive comment helpful. It's not clear to me what this is doing or why it's packed like it is.

banach-space replied:
Sadly I won't have the bandwidth to expand just now :( I will try to once I get to scalable vectorisation.

c-rhodes reviewed (approved):
I've left some comments mostly minor, I'm accepting this on the basis it demonstrates a pack + mmt4d + unpack is equivalent to a matmul, which is easily verifiable from the output.

If the purpose of this test is also for documentation I think it could be improved, particularly with some comments to clarify the packing, but additionally I think a simpler example with smaller shapes would be better.

banach-space commented on mlir/test/Integration/Dialect/Linalg/CPU/pack-unpack-mmt4d.mlir:
```diff
@@ -0,0 +1,174 @@
+// DEFINE: %{compile} =  mlir-opt %s \
+// DEFINE:    -transform-interpreter -test-transform-dialect-erase-schedule \
+// DEFINE:    -one-shot-bufferize="bufferize-function-boundaries" -test-lower-to-llvm -o %t -o %t
```
This one will hunt me for a while! Thanks, great catch 🙏🏻

banach-space committed:
fixup! fixup! [mlir][linalg] Add e2e test for linalg.mmt4d

Address PR comments

c-rhodes commented on mlir/test/Integration/Dialect/Linalg/CPU/pack-unpack-mmt4d.mlir:
```diff
@@ -0,0 +1,172 @@
+// DEFINE: %{compile} =  mlir-opt %s \
+// DEFINE:    -transform-interpreter -test-transform-dialect-erase-schedule \
+// DEFINE:    -one-shot-bufferize="bufferize-function-boundaries" \
+// DEFINE:    -buffer-deallocation-pipeline  -cse -canonicalize -convert-vector-to-scf -test-lower-to-llvm
```
`-convert-vector-to-scf` can be removed, it's the first pass in `-test-lower-to-llvm`. Also, replace double space before `-cse` with single space.

```suggestion
// DEFINE:    -buffer-deallocation-pipeline -cse -canonicalize -test-lower-to-llvm
```

banach-space committed:
fixup! fixup! fixup! [mlir][linalg] Add e2e test for linalg.mmt4d

Add "private-function-dynamic-ownership" to
"-buffer-deallocation-pipeline". This prevents the bufferizer from
inserting `bufferization.clone` (which are a bit tricky to remove) while
still avoiding leaks.

bjacob commented:
This looks great, thanks for doing this! Not formally participating in the review process as I'm unfamiliar with transform dialect.
